### DISCUSSION TITLE
13:38 data type issues in training loop - Thanks for the great videos!

### DISCUSSION THREAD

**Discussion by JackXu9** [mrdbourke/pytorch-deep-learning#86] [Q&A] [answered]
Hi, I've been struggling a bit with fixing one of the data type issues in the training loop related to the loss function (in the multi class classification model). I receive the following Runtimeerror:

`RuntimeError: "log_softmax_lastdim_kernel_impl" not implemented for 'Long'`

I have tried to match the input and target types for the loss function within the training loop (int64). I did this because when I changed the initial data conversion to tensors from `torch.float` to `torch.LongTensor`, I ran into more type issues prior to the training loop. 
The code for the loop is as follows:
![image](https://user-images.githubusercontent.com/94681350/183290427-2061fd45-242f-4d5c-bb03-c3c4a81a5dbd.png)

**Comment by aronvandepol** [comment, marked as answer]
Hi Jack, 

Have you tried changing the tensor type to `torch.float32`? I managed to get through almost all with simply that datatype.

You can also see the error implies that you cannot call the loss function on a `torch.LongTensor`.

**Reply by JackXu9** [reply]
Hi Aron,

thanks for the tip.
So now at least it is running with float32, however the test results seems a bit off.. XD
I did hit a few other errors after the type change, such as `Runtime Error - element 0 of tensors does not require grad and does not have a grad_fn`. I fixed that with `loss.requires_grad = True` in the training loop. But I'm not sure what is happening with my loss and acc calculations.
![image](https://user-images.githubusercontent.com/94681350/183303564-635404f0-13f5-49bd-84de-5566f564b91c.png)

**Reply by aronvandepol** [reply]
Hi Jack,

After a careful look, I think I see where your problem lies. You are calling `torch.softmax().argmax()` on your forward_pass data (your y_logits). You want to use the logits to calculate your loss function, and not the prediction labels. Using  `torch.softmax()` and `.argmax()` gives your prediction labels. if you calculate your predicition labels only when needed, you should be fine.

To conclude.

- Using `torch.softmax()` on your multi-class models **raw outputs** gives you the **prediction probabilities** for each class in tensor format. 
- Calling `.argmax()` on your **prediction probabilities** tensor gets you your **prediction labels** (e.g class: 1 or 2). 
- You want to use your models **raw output (logits)** in your loss function and not **predictions probabilities** or **prediction labels**.

Hope this helps,
Aron.

**Reply by JackXu9** [reply]
Ahh yes of course. I also had to change the y_blob to LongTensor as in the video, and then it worked. I just thought that the loss function required the same datatype for prediction and target. But thanks for the quick reply!
![image](https://user-images.githubusercontent.com/94681350/183394502-074ba729-6a9d-4c93-b214-0dbad5df0681.png)

**Reply by aronvandepol** [reply]
To make sure, I believe they do require the same datatype, and if not; the main issue just lied with the fact that you called `softmax()` and `argmax()`. 

Also for next time, instead of screenshotting your code, copy it and put it in in between two grave icons \`YOUR CODE IN BETWEEN\` and renders out as: `YOUR CODE IN BETWEEN` making it easier for us to check your code.

Anyhow, happy it worked out and your model is training.

**Reply by JackXu9** [reply]
How does new line work in the `code function`? I can't seem to get it to create new lines with \n or ENTER.

**Reply by aronvandepol** [reply]
you can use 

\```
your code
hello world
\```

renders as:

```
your code
hello world
```

**Comment by aronvandepol** [comment]
@mrdbourke this one is answered and solved :)

**Reply by mrdbourke** [reply]
Thank you so much for your help! :)
